File: 백준/Gold/2565. 전깃줄/전깃줄.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	N = Integer.parseInt(br.readLine());
    	arr = new Pair[N+1];
    	dp = new int[N+1];
    	for(int i = 1; i <= N; i++) {
    		st = new StringTokenizer(br.readLine());
    		int a = Integer.parseInt(st.nextToken());
    		int b = Integer.parseInt(st.nextToken());
    		
    		arr[i] = new Pair(a,b);
    	}
    	
    	Arrays.sort(arr, 1, N+1);
    	
    	
    	for(int i = 1; i<=N;i++) {
    		ans = Math.max(ans, solve(i));
    	}
    	System.out.println(N - ans);
    }
    
    static int solve(int cur) {
    	if(dp[cur] != 0)
    		return dp[cur];
    	int ret = 1;
    	
    	for(int i = 1; i< cur; i++) {
    		if(check(arr[i], arr[cur]))
    			ret = Math.max(ret, solve(i) + 1);
    	}
    	
    	return dp[cur] = ret;
    }
    
    static boolean check(Pair prev, Pair cur) {
    	return prev.b < cur.b;
    }
    
    static int dp[];
    
    static int ans;
    
    static int N;
    static StringTokenizer st;
    
    static Pair arr[];
    
    static class Pair implements Comparable<Pair>{
    	int a, b;
    	public Pair(int a, int b) {
    		this.a = a;
    		this.b = b;
    				
    	}
    	
    	public int compareTo(Pair o) {
    		if( a != o.a)
    			return a - o.a;
    		return b - o.b;
    	}
    }
    
}

File: 백준/Gold/2565. 전깃줄/전깃줄.py
from collections import namedtuple

def solve(cur, last):
    if cur == N + 1:
        return 0

    if dp[cur][last] != -1:
        return  dp[cur][last]

    ret = solve(cur + 1, last)

    if arr[cur].b > arr[last].b:
        ret = max(ret, solve(cur+1, cur) + 1)

    dp[cur][last] = ret
    return ret

if __name__ == '__main__':
    N = int(input())

    Pair = namedtuple('Pair', ['a','b'])
    INF = int(1e10)
    arr = []
    for i in range(N):
        a,b = map(int,input().split())
        arr.append(Pair(a,b))

    arr = [Pair(0,-INF)] + arr

    arr.sort(key=lambda pair : pair.a)
    dp = [[-1] * (N+1) for _ in range(N+1)]
    print(N - solve(1, 0))